utils/wasm-gen/src/utils.rs

```diff
@@ -215,6 +222,172 @@ fn find_recursion_impl<Callback>(
+                    .build(),
+            );
+
+            // back to plain module
+            let module = mbuilder.build();
+
+            let import_count = module.import_count(ImportCountType::Function);
+            let gr_gas_available_index = import_count as u32 - 1;
+
+            (gr_gas_available_index, module)
+        }
+    };
+
+    let (Some(type_section), Some(function_section)) =
+        (module.type_section(), module.function_section())
+    else {
+        return module;
+    };
+
+    let types = type_section.types().to_vec();
+    let signature_entries = function_section.entries().to_vec();
+
+    let Some(code_section) = module.code_section_mut() else {
+        return module;
+    };
+
+    for (index, func_body) in code_section.bodies_mut().iter_mut().enumerate() {
+        let signature_index = &signature_entries[index];
+        let signature = &types[signature_index.type_ref() as usize];
+        let Type::Function(signature) = signature;
+        let results = signature.results();
+
+        let mut body = Vec::with_capacity(results.len() + 9);
+        body.extend_from_slice(&[
+            Instruction::I32Const(gas_ptr as i32),
+            Instruction::Call(gr_gas_available_index),
+            Instruction::I32Const(gas_ptr as i32),
+            Instruction::I64Load(3, 0),
+            Instruction::I64Const(critical_gas_limit as i64),
+            Instruction::I64LeU,
+            Instruction::If(BlockType::NoResult),
+        ]);
+
+        for result in results {
+            let instruction = match result {
+                ValueType::I32 => Instruction::I32Const(u32::MAX as i32),
+                ValueType::I64 => Instruction::I64Const(u64::MAX as i64),
+                ValueType::F32 | ValueType::F64 => unreachable!("f32/64 types are not supported"),
+            };
+
+            body.push(instruction);
+        }
+
+        body.extend_from_slice(&[Instruction::Return, Instruction::End]);
+
+        let instructions = func_body.code_mut().elements_mut();
+
+        let original_instructions =
+            mem::replace(instructions, Vec::with_capacity(instructions.len()));
+        let new_instructions = instructions;
+
+        new_instructions.extend_from_slice(&body);
+
+        for instruction in original_instructions {
+            match instruction {
+                Instruction::Block(_) | Instruction::Loop(_) | Instruction::If(_) => {
+                    new_instructions.push(instruction);
+                    new_instructions.extend_from_slice(&body);
+                }
+                Instruction::Call(call_index)
+                    if rewrite_calls && call_index >= gr_gas_available_index =>
+                {
+                    new_instructions.push(Instruction::Call(call_index + 1));
+                }
+                _ => {
+                    new_instructions.push(instruction);
+                }
+            }
+        }
+    }
+
+    if rewrite_calls {
```

Comment:
Arent we able to push call back?

Reply:
Not sure what you mean. We need to rewrite the sections after inserting the imports, just like the wasm-instrument does: https://github.com/gear-tech/wasm-instrument/blob/v0.2.1-sign-ext/src/gas_metering/mod.rs#L235